### PR TITLE
Relax path validation (#2355)

### LINKED ISSUE #2349: Accessing a file with a percent-encoded name on the filesystem with ObjectStore LocalFileSystem
**Describe the bug**

The issue was discovered when trying to read a percent-encoded filename using DataFusion.

Upon investigation, the issue maybe partly in `LocalFileSystem` function `path_to_filesystem`: this function will percent-encoding the path when calling `extend` on `parts`.

As such, a file such as `L%3ABC.parquet` cannot be accessed by `LocalFileSystem`.

`L%3ABC.parquet` is the actual filename on the file system as reported by the command line `ls`, for example.

Note: `L%3ABC` is the percentage-encoded version of `L:BC`.

**To Reproduce**

Create a data file (not using `LocalFileSystem`) named `L%3ABC.parquet`. Creating the file on the file system outside of `LocalFileSystem` is easiest. This file could be created from tools like Python/Pandas, a database, etc.

Try reading it with datafusion as such:
```
{
    let options = ParquetReadOptions::default();
    let fs_name = "L%3ABC.parquet"; // this is the name on the file system
    let r = ctx.read_parquet(fs_name, options).await.unwrap();
}
```

This will fail with an error similar to (added newlines):

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value:
ObjectStore(NotFound { path: "...../data/L%253ABC.parquet",
   source: Os { code: 2, kind: NotFound, message: "No such file or directory" } })', src/main.rs:88:62
```

Trying other variation does not help: `L:BC.parquet` errors else where.

If the file is named on the filesystem `L%253ABC.parquet`, one can read the file with:
```
let fs_name = "L%3ABC.parquet"; // this will read L%253ABC.parquet 
let r = ctx.read_parquet(fs_name, options).await.unwrap();
```

**Expected behavior**

The file `L%3ABC.parquet` should be read by the context.

**Additional context**

After some investigation, there might be some double encoding going on in Object Store's `LocalFileSystem`: in `Config:path_to_filesystem` here: https://github.com/apache/arrow-rs/blob/38764c26db511ea13538042f229e817562f02f74/object_store/src/local.rs#L228

More specifically `extend()` will percent-encode the string again: https://github.com/apache/arrow-rs/blob/38764c26db511ea13538042f229e817562f02f74/object_store/src/local.rs#L235

I worked out a tentative patch, but I came across this "illegal" path test:
https://github.com/apache/arrow-rs/blob/38764c26db511ea13538042f229e817562f02f74/object_store/src/local.rs#L1200

the patch was able to access such file and failed on the test:
https://github.com/apache/arrow-rs/blob/38764c26db511ea13538042f229e817562f02f74/object_store/src/local.rs#L1211 

The patch was able to handle this kind of filenames (vs "Path").
Clarification on this kind of path would be helpful to adjust the patch or change the test?

**Tests**

Below are some "cases" to show the behaviour.
The additional logs `seg = ...` come from the function `path_to_filesystem` - I added a simple log for each segments.

https://github.com/apache/arrow-rs/blob/38764c26db511ea13538042f229e817562f02f74/object_store/src/local.rs#L228

Case 1 and 2 fail; but case 3 succeeds.

***Case 1***

File present:

```
L%3ABC.parquet
```

Code:
```
    let ctx = SessionContext::new();
    let options = ParquetReadOptions::default();
    let fs_name = "data/L%3ABC.parquet"; // this is as on FS
    let r = ctx.read_parquet(fs_name, options).await.unwrap();
```

Result:
```
==================================================
        seg = PathPart { raw: "Users" }
        seg = PathPart { raw: "jc" }
        seg = PathPart { raw: "src" }
        seg = PathPart { raw: "rust" }
        seg = PathPart { raw: "timeseries" }
        seg = PathPart { raw: "data" }
        seg = PathPart { raw: "L%253ABC.parquet" }
url = Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/Users/jc/src/rust/timeseries/data/L%25253ABC.parquet", query: None, fragment: None }
res = Ok("/Users/jc/src/rust/timeseries/data/L%253ABC.parquet")
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ObjectStore(NotFound { path: "/Users/jc/src/rust/timeseries/data/L%253ABC.parquet", source: Os { code: 2, kind: NotFound, message: "No such file or directory" } })', src/main.rs:85:54
```

***Case 2***

File present:

```
L%253ABC.parquet
```

Code:
```
    let ctx = SessionContext::new();
    let options = ParquetReadOptions::default();
    let fs_name = "data/L%3ABC.parquet"; // this is as on FS
    let r = ctx.read_parquet(fs_name, options).await.unwrap();
```

Result:
```
==================================================
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Os { code: 2, kind: NotFound, message: "No such file or directory" })', src/main.rs:85:54
```

***Case 3***

Files present (yes I had to have the 2 files on disk to make this work):
```
L%253ABC.parquet
L%3ABC.parquet
```

```
    let ctx = SessionContext::new();
    let options = ParquetReadOptions::default();
    let fs_name = "data/L%3ABC.parquet"; // this is as on FS
    let r = ctx.read_parquet(fs_name, options).await.unwrap();
```

Result:
```
==================================================
        seg = PathPart { raw: "Users" }
        seg = PathPart { raw: "jc" }
        seg = PathPart { raw: "src" }
        seg = PathPart { raw: "rust" }
        seg = PathPart { raw: "timeseries" }
        seg = PathPart { raw: "data" }
        seg = PathPart { raw: "L%253ABC.parquet" }
url = Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/Users/jc/src/rust/timeseries/data/L%25253ABC.parquet", query: None, fragment: None }
res = Ok("/Users/jc/src/rust/timeseries/data/L%253ABC.parquet")
        seg = PathPart { raw: "Users" }
        seg = PathPart { raw: "jc" }
        seg = PathPart { raw: "src" }
        seg = PathPart { raw: "rust" }
        seg = PathPart { raw: "timeseries" }
        seg = PathPart { raw: "data" }
        seg = PathPart { raw: "L%253ABC.parquet" }
url = Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/Users/jc/src/rust/timeseries/data/L%25253ABC.parquet", query: None, fragment: None }
res = Ok("/Users/jc/src/rust/timeseries/data/L%253ABC.parquet")
        seg = PathPart { raw: "Users" }
        seg = PathPart { raw: "jc" }
        seg = PathPart { raw: "src" }
        seg = PathPart { raw: "rust" }
        seg = PathPart { raw: "timeseries" }
        seg = PathPart { raw: "data" }
        seg = PathPart { raw: "L%253ABC.parquet" }
url = Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/Users/jc/src/rust/timeseries/data/L%25253ABC.parquet", query: None, fragment: None }
res = Ok("/Users/jc/src/rust/timeseries/data/L%253ABC.parquet")
````

### LINKED ISSUE #2355: Relax Path Validation to Allow Any Percent-Encoded Sequence
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The current validation performed by `PathPart::parse` and by extension `Path::parse` returns an error if the provided string does not round-trip. In particular it decodes the string, re-encodes it, and if the result does not equal the original string, it returns an error.

This is safe, but can lead to surprises when interacting with other systems. In particular it will return an error for:

* Percent-encoded sequences for characters `object_store` does not consider illegal #2349 
* Paths that contain a `%` but not as part of a percent encoded sequence

**Describe the solution you'd like**

There is no reason not to permit these paths, especially in the former case where they are arguably more safe than the paths generated by `object_store`. To phrase it differently `Path::parse` should accept any "safe" path, even if it would not itself generate that path.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2355
Closes #2349 

# Rationale for this change
 

See tickets

# What changes are included in this PR?

This relaxes the validation in `PathPart::parse` to be less conservative about what constitutes a safe path. In particular it should accept any "safe" path, even if it would not generate such a path itself.

# Are there any user-facing changes?

No

### CODE REVIEW

File: object_store/src/local.rs [hunk 1]
```diff
@@ -1247,4 +1247,23 @@ mod tests {
             0
         );
     }
+
+    #[tokio::test]
+    async fn filesystem_filename_with_percent() {
```
Comment (tustvold) [inline object_store/src/local.rs:1252]: This is @jccampagne 's test case from #2351

File: object_store/src/path/parts.rs [hunk 2]
```diff
@@ -43,21 +47,39 @@ pub struct PathPart<'a> {
 impl<'a> PathPart<'a> {
     /// Parse the provided path segment as a [`PathPart`] returning an error if invalid
     pub fn parse(segment: &'a str) -> Result<Self, InvalidPart> {
-        let decoded: Cow<'a, [u8]> = percent_decode(segment.as_bytes()).into();
-        let part = PathPart::from(decoded.as_ref());
-        if segment != part.as_ref() {
+        if segment == "." || segment == ".." {
```
Comment (crepererum) [inline object_store/src/path/parts.rs:50]: Could some `percent_validate` be upstreamed to `percent_encoding`?

Comment (tustvold) [reply object_store/src/path/parts.rs:50]: It would certainly be helpful if AsciiSet had a public interface to check a byte

**Review by crepererum — APPROVED** [review]
Upstreaming would be nice, otherwise this looks good.

File: object_store/src/path/parts.rs (hunk 2 above)
Comment (tustvold) [reply object_store/src/path/parts.rs:50]: Filed a ticket https://github.com/servo/rust-url/issues/784

**Pull request merged**